Add Rabbitmq support

amir-mhp opened the pull request:
<!--  
Thanks for sending a pull request!  Here are some tips for you: 
感谢你提交了 pull request ，一些 tips 供您参考：
-->

**What type of PR is this?**
/kind feature


**What this PR does / why we need it**:
**这个PR解决了什么问题**:

This PR adds RabbitMQ support to the K8s Eventer module. RabbitMQ is a widely used message broker that enables efficient communication between different components in a distributed system. By integrating RabbitMQ support, the K8s Eventer module can now leverage RabbitMQ's features for reliable message delivery, scalability, and decoupling of components.

**Which issue(s) this PR fixes**:
**PR相关联issue**:

Fixes #


**Does this PR introduce a breaking change?**:
**PR带来的破坏性变更**:

```docs

```

**Test/Final result**:
**测试/最终运行结果**:

I have conducted the following tests to ensure the functionality of the RabbitMQ support:

Integration test with RabbitMQ messaging.
Compatibility test with existing Kubernetes Eventer features.
Performance test to evaluate the efficiency of RabbitMQ communication.

Results:
All tests passed successfully, demonstrating the robustness and compatibility of the RabbitMQ integration.

Commit:
Init rabbitmq on common

Commit:
Add glog adapter for rabbitmq

Commit:
Init rabbitmq client

Commit:
Add rabbitmq dependency to go

Commit:
Update vendors

Commit:
Add Dialer and tls to rabbit config

Commit:
Add produce amqp message logic

Commit:
Add rabbitmq driver

Commit:
Add rabbitmq driver test

Commit:
Add rabbitmq into factory build

Commit:
Minor changes

Commit:
Add testify dependency

Commit:
Fix Get rabbitmq url

Commit:
Add rabbitmq into docs

Commit:
Minor changes at docs

amir-mhp committed:
Merge branch 'master' into master

amir-mhp commented:
Hi @KeyOfSpectator  maintainers

I hope you're doing well. Just wanted to kindly follow up on this PR: https://github.com/AliyunContainerService/kube-eventer/pull/284 — which adds RabbitMQ support to the K8s Eventer module.

This enhancement enables better decoupling and scalability by integrating a widely adopted message broker, and it has passed all integration, compatibility, and performance tests successfully.

I understand maintainers may be busy or the PR might have been missed, but I’d appreciate it if you could take a moment to review it when possible — or let me know if there are any blockers, suggestions, or required changes.

Thanks for your time and for maintaining this awesome project! 🙏

Best regards,
@amir-mhp

yshngg commented on common/rabbitmq/glogadapter.go:
```diff
@@ -0,0 +1,27 @@
+	_ "k8s.io/klog"
+)
+
+type Logging interface {
+	Print(v ...interface{})
+	Printf(format string, v ...interface{})
+	Println( v ...interface{})
+}
```
What is the purpose of defining the `Logger` interface here? It seems somewhat redundant.

The `Logger` interface in the `github.com/rabbitmq/amqp091-go` package is defined as:

```go
type Logging interface {
	Printf(format string, v ...interface{})
}
```

Thus, only the `Printf` method needs to be implemented by the `GologAdapterLogger` struct.

Ref: https://github.com/rabbitmq/amqp091-go/blob/5eb51bef315be9a8721bb396540a5765df38bf0e/log.go#L7-L9

yshngg commented on common/rabbitmq/rabbitmq.go:
```diff
@@ -0,0 +1,239 @@
+
+	amqp.Logger = GologAdapterLogger{}
+
+	TLSClientConfig, TLSClientConfigEnable, err := getTlsConfiguration(opts)
```
Missing error handling.

KeyOfSpectator commented:
nice feature

Commit:
fix(rabbitmq): remove redundant logging interface and add error handling

- Remove custom Logging interface in favor of amqp091.Logging interface
- Implement only Printf method as required by AMQP library
- Add missing error handling for getTlsConfiguration call

amir-mhp commented:
@yshngg Thanks for the thorough review! I've addressed all the issues you raised:

## Fixed Issues:

1. **Duplicate package import** in `glogadapter.go`: Removed the duplicate `k8s.io/klog` import
2. **Redundant logging interface**: Removed custom interface and now only implement `Printf` method as required by amqp091
3. **Missing error handling**: Added proper error handling for `getTlsConfiguration` call
4. **Indirect dependency comment**: Removed `// indirect` from `github.com/rabbitmq/amqp091-go` in go.mod

## Regarding testify upgrade:
The `github.com/stretchr/testify` upgrade from v1.6.1 to v1.8.0 was likely pulled in automatically when adding the RabbitMQ dependency. However, I can revert this if it's not needed for this feature. Should I:
- Keep the upgrade (brings bug fixes and improvements)
- Revert to v1.6.1 to minimize changes
- Or is there a specific version requirement for this project?

All changes have been pushed and conflicts resolved. Ready for re-review!

@KeyOfSpectator Thanks for the positive feedback

yshngg commented:
Good job @amir-mhp !

I've validated locally - as you mentioned, adding the RabbitMQ dependency automatically does upgrade `github.com/stretchr/testify` from **v1.6.1** to **v1.8.0**.

This occurs due to the following dependency chain:

```console
github.com/rabbitmq/amqp091-go@v1.9.0 -> go.uber.org/goleak@v1.2.1 -> github.com/stretchr/testify@v1.8.0
```
We should keep the upgraded version (**v1.8.0**). Could you please adjust the changes accordingly?

Thanks for contributing! 🎉

Commit:
Upgrade testify to 1.8.0

amir-mhp commented:
@yshngg Thank you for validating that locally and explaining the dependency chain! 🙏 

That makes perfect sense - the testify upgrade is indeed automatic due to:
`rabbitmq/amqp091-go@v1.9.0 → go.uber.org/goleak@v1.2.1 → stretchr/testify@v1.8.0`

I'll keep the v1.8.0 upgrade as recommended. All the code changes are already in place and working correctly.

Thanks for the thorough review and for taking the time to validate the dependency chain locally. Really appreciate the detailed feedback! 

Ready to merge when you are! 🚀

yshngg reviewed (approved):
lgtm

amir-mhp commented:
Hi @KeyOfSpectator 👋

The RabbitMQ support PR (#284) has now been fully reviewed, approved, and all requested changes have been addressed. ✅
Thanks again for your kind feedback earlier — just wanted to kindly follow up and see if you could proceed with the merge when you have a moment. 🚀

Let me know if there’s anything else I can help with.

Best regards,
@amir-mhp 🙏

amir-mhp commented:
Hi @KeyOfSpectator  👋

The RabbitMQ support PR (https://github.com/AliyunContainerService/kube-eventer/pull/284) has now been fully reviewed, approved, and all requested changes have been addressed. ✅
Thanks again for your kind feedback earlier — just wanted to kindly follow up and see if you could proceed with the merge when you have a moment. 🚀

Let me know if there’s anything else I can help with.

Best regards,
@amir-mhp  🙏

amir-mhp commented:
Hi @KeyOfSpectator maintainers

I hope you're doing well. Just wanted to kindly follow up on this PR: https://github.com/AliyunContainerService/kube-eventer/pull/284 — which adds RabbitMQ support to the K8s Eventer module.

This enhancement enables better decoupling and scalability by integrating a widely adopted message broker, and it has passed all integration, compatibility, and performance tests successfully.

I understand maintainers may be busy or the PR might have been missed, but I’d appreciate it if you could take a moment to review it when possible — or let me know if there are any blockers, suggestions, or required changes.

Thanks for your time and for maintaining this awesome project! 🙏

Best regards,
@amir-mhp
